### PR TITLE
Refactor ExecutionCourse getAttends methods

### PR DESCRIPTION
**Pull request by pedromonteiro-qubIT**
Relates with: #qubIT-Fenix-8842

### CODE REVIEW

**Comment by pedromonteiro-qubIT** [comment]
## Tests for `getAttendsByStudent`

### `Registration` overload

| Test | Expected behavior |
|------|-------------------|
| `matchFound` | Finds the `Attends` for a `Registration` enrolled in the course |
| `noMatch` | Returns `null` when the `Registration` is not enrolled |
| `emptySet` | Returns `null` when the course has no `Attends` |
| `filtersAmongMultipleAttends` | Picks the correct `Attends` among different students |
| `identitySemantics` | Returns the right `Attends` for one `Registration` and `null` for another |
| `usesIdentityNotEquality` | Lookup uses identity, not equality |
| `multipleMatches` | When two `Attends` exist for the same `Registration`, returns one of them |

### `Student` overload

| Test | Expected behavior |
|------|-------------------|
| `matchFound` | Finds the `Attends` via a `Student` that is enrolled |
| `noMatch` | Returns `null` when the `Student` is not enrolled |
| `emptySet` | Returns `null` when the course has no `Attends` |
| `multipleMatches` | When two `Attends` exist for the same `Student`, returns one of them |
| `withStudentWithoutRegistrations_returnsNull` | A `Student` with no `Registration` gets `null`, even if others have `Attends` |

### `null` handling

| Test | Expected behavior |
|------|-------------------|
| `withNullRegistration` | Passing `null` as `Registration` returns `null` |
| `withNullStudent` | Passing `null` as `Student` returns `null` |

### Overload coherence

| Test | Expected behavior |
|------|-------------------|
| `overloadCoherence` | Both overloads (`Registration` and `Student`) return the same `Attends` |

File: src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    private static Registration regB;
+    private ExecutionCourse ec;
+
+    @BeforeClass
+    public static void init() {
+        FenixFramework.getTransactionManager().withTransaction(() -> {
+            EnrolmentTest.initEnrolments();
+
+            registration = Student.readStudentByNumber(1).getRegistrationStream().findAny().orElseThrow();
+            student = registration.getStudent();
```
Comment (josetrigo-qubit) [inline src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java]: student = Student.readStudentByNumber(1);
registration = student.getRegistrationStream().findAny().orElseThrow();

File: src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+            executionCourse = registration.getAssociatedAttendsSet().iterator().next().getExecutionCourse();
+
+            differentReg = createNewRegistration("Different", "different." + UUID.randomUUID());
```
Comment (josetrigo-qubit) [inline src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java]: nonAttendingRegistration would be a better name

File: src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+    @Before
+    public void setUp() {
+        ec = createEmptyExecutionCourse();
+    }
+
+    private static ExecutionCourse createEmptyExecutionCourse() {
+        final String uuid = UUID.randomUUID().toString();
+        final ExecutionInterval interval = ExecutionInterval.findFirstCurrentChild(
+                registration.getLastStudentCurricularPlan().getDegree().getCalendar());
```
Comment (josetrigo-qubit) [inline src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java]: You can use ExecutionInterval.findFirstCurrentChild(null)

File: src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        return new ExecutionCourse(uuid, uuid, interval);
+    }
+
+    private static Registration createNewRegistration(final String name, final String username) {
+        final Student s = StudentTest.createStudent(name, username);
+        final Degree degree = Degree.find(DEGREE_A_CODE);
+        assertNotNull(degree);
+        final DegreeCurricularPlan dcp = degree.getDegreeCurricularPlansSet().stream()
+                .filter(p -> DCP_NAME_V1.equals(p.getName())).findAny().orElseThrow();
+        final ExecutionYear executionYear = ExecutionYear.findCurrent(dcp.getDegree().getCalendar());
```
Comment (josetrigo-qubit) [inline src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java]: You can use ExecutionYear.findCurrent(null)

File: src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+            regA = createNewRegistration("RegA", "reg.a." + UUID.randomUUID());
+            regB = createNewRegistration("RegB", "reg.b." + UUID.randomUUID());
+
+            return null;
+        });
+    }
+
+    @Before
+    public void setUp() {
+        ec = createEmptyExecutionCourse();
```
Comment (josetrigo-qubit) [inline src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java]: You should give variables more descriptive name to reduce mental load while reading the tests like emptyExecutionCourse

File: src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+    @Test
+    public void getAttendsByStudent_withRegistration_filtersAmongMultipleAttends() {
+        final Attends a = new Attends(regA, ec);
+        new Attends(regB, ec);
+        assertSame(a, ec.getAttendsByStudent(regA));
+    }
+
+    @Test
+    public void getAttendsByStudent_withRegistration_identitySemantics() {
```
Comment (josetrigo-qubit) [inline src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java]: This test basically combines getAttendsByStudent_withRegistration_matchFound + getAttendsByStudent_withRegistration_noMatch. Why is it needed?

File: src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        final Attends attends = new Attends(registration, ec);
+        assertSame(attends, ec.getAttendsByStudent(registration));
+        assertNull(ec.getAttendsByStudent(differentReg));
+    }
+
+    @Test
+    public void getAttendsByStudent_withRegistration_usesIdentityNotEquality() {
```
Comment (josetrigo-qubit) [inline src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java]: How is this test different from getAttendsByStudent_withRegistration_identitySemantics?

File: src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        assertSame(byRegistration, byStudent);
+    }
+
+    @Test
+    public void getAttendsByStudent_withRegistration_multipleMatches() {
+        final Attends a1 = new Attends(registration, ec);
+        final Attends a2 = new Attends(registration, ec);
+
+        assertNotSame(a1, a2);
+        assertEquals(2, ec.getAttendsSet().size());
```
Comment (josetrigo-qubit) [inline src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java]: Is this needed? We are testing getAttendsByStudent, not Attends constructor. This comment also applies for getAttendsByStudent_withStudent_multipleMatches

File: src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+        final Attends result = ec.getAttendsByStudent(registration);
+        assertNotNull(result);
```
Comment (josetrigo-qubit) [inline src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java]: assertNotNull(result) is not needed. We already tested getAttendsByStudent()'s behaviour in getAttendsByStudent_withRegistration_matchFound. We should focus on the test case we are actually trying to test: multipleMatches

File: src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    public static void init() {
+        FenixFramework.getTransactionManager().withTransaction(() -> {
+            EnrolmentTest.initEnrolments();
+
+            registration = Student.readStudentByNumber(1).getRegistrationStream().findAny().orElseThrow();
+            student = registration.getStudent();
+            executionCourse = registration.getAssociatedAttendsSet().iterator().next().getExecutionCourse();
+
+            differentReg = createNewRegistration("Different", "different." + UUID.randomUUID());
+            otherStudent = StudentTest.createStudent("Other", "other." + UUID.randomUUID());
```
Comment (josetrigo-qubit) [inline src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java]: studentWithoutRegistrations would be a better name

File: src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+        assertNull(executionCourse.getAttendsByStudent((Student) null));
+    }
+
+    @Test
+    public void getAttendsByStudent_overloadCoherence() {
+        final Attends byRegistration = executionCourse.getAttendsByStudent(registration);
+        final Attends byStudent = executionCourse.getAttendsByStudent(student);
+        assertNotNull(byRegistration);
+        assertNotNull(byStudent);
+        assertSame(byRegistration, byStudent);
```
Comment (josetrigo-qubit) [inline src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java]: This only passes because we assume the student has a single registration. If a second registration were added later, getAttendsByStudent(student) could return a different Attends than getAttendsByStudent(registration) because of findFirst().

Also, If we have an overload coherence test that confirms both overloads behave the same, do we need duplicate test cases for every scenario? Once coherence is verified, we can test the behavior once and trust the other overload follows suit. This would cut the test file size significantly. For example, noMatch, emptySet, null, and multipleMatches each appear twice with near-identical logic.

**Review by josetrigo-qubit — CHANGES_REQUESTED** [review]

File: src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java (hunk 7 above)
Comment (pedromonteiro-qubIT) [reply src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java]: Good catch! 

I initially thought of splitting the **same registration** and **different registration** scenarios into separate tests, but `identitySemantics` already covers both in one method. `usesIdentityNotEquality` ends up testing the exact same thing with different fixtures, adding no real insight about identity vs equality. It is indeed redundant.

 I'll remove it.

File: src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java (hunk 11 above)
Comment (pedromonteiro-qubIT) [reply src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java]: In the current setup, student has exactly one registration (`initStudentAndRegistration` creates it with a single `createRegistration` call, and nothing else in the chain creates another). So divergence between the overloads can't happen. It would require two registrations for the same student in the same execution course.

That said, that guarantee is invisible at the test site. So instead of relying on it, I tied the coherence to a locally created student. Since **regA** comes from `createNewRegistration` and `createStudent`, its student is created for that single registration, so coherence now holds by construction and the second-registration hypothesis can no longer touch the test:

```java
@Test
public void getAttendsByStudent_overloadCoherence() {
    final Attends attends = new Attends(regA, emptyExecutionCourse);
    final Student student = regA.getStudent();
    assertSame(attends, emptyExecutionCourse.getAttendsByStudent(regA));
    assertSame(attends, emptyExecutionCourse.getAttendsByStudent(student));
}
```

**Review by josetrigo-qubit — APPROVED** [review]

File: src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+        assertSame(attends, emptyExecutionCourse.getAttendsByStudent(student));
+    }
+
+    @Test
+    public void getAttendsByStudent_withRegistration_multipleMatches() {
+        final Attends a1 = new Attends(registration, emptyExecutionCourse);
+        final Attends a2 = new Attends(registration, emptyExecutionCourse);
+
+        final Attends result = emptyExecutionCourse.getAttendsByStudent(registration);
+        assertTrue(Objects.equals(result, a1) || Objects.equals(result, a2));
```
Comment (sanavarali) [inline src/test/java/org/fenixedu/academic/domain/ExecutionCourseTest.java]: This assert is strange, as it is non-deterministic. Anyway this should never happen because the Attends constructor is package protected and the ony way to create it's from `Enrolment.findOrCreateAttends(ExecutionCourse)`, so this test can be deleted.

**Review by sanavarali — CHANGES_REQUESTED** [review]

**Review by sanavarali — APPROVED** [review]

**Review by Sofia-Nascimento — APPROVED** [review]
